sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -0,0 +1,1191 @@
+----
+|This function adds a command group function object to a graph. The function
+object statically contains a group of commands, of which a single command is
+executed at runtime. A function object can be a host task which is scheduled by
+the SYCL runtime, or a SYCL function for invoking kernels with all restrictions
+that apply as described in the core specification. The requisites of `cgf` will
+be used to identify any dependent nodes in the graph to form edges with.
+
+Preconditions:
+
+* This member function is only available when the `command_graph` state is
+  `graph_state::modifiable`.
+
+Parameters:
+
+* `cgf` - Command group function object to be added as a node.
+
+* `propList` - Zero or more properties can be provided to the constructed node
+  via an instance of `property_list`.
+
+Returns: The command-group function object node which has been added to the graph.
+
+Exceptions:
+
+* Throws synchronously with error code `invalid` if a queue is recording
+  commands to the graph.
+
+|
+[source,c++]
+----
+using namespace ext::oneapi::experimental;
+void make_edge(node src, node dest);
+----
+
+|Creates a dependency between two nodes representing a happens-before relationship.
+
+Preconditions:
+
+* This member function is only available when the `command_graph` state is
+  `graph_state::modifiable`.
+
+Parameters:
+
+* `src` - Node which will be a dependency of `dest`.
+
+* `dest` - Node which will be dependent on `src`.
+
+Exceptions:
+
+* Throws synchronously with error code `invalid` if a queue is recording
+  commands to the graph object.
+
+* Throws synchronously with error code `invalid` if `src` or `dest`
+  are not valid nodes assigned to the graph object.
+
+* Throws synchronously with error code `invalid` if `src` and `dest`
+  are the same node.
+
+|
+[source,c++]
+----
+using namespace ext::oneapi::experimental;
+command_graph<graph_state::executable> finalize(const context& syclContext) const;
+----
+
+|Synchronous operation that creates a new graph in the executable state with a
+fixed topology that can be submitted for execution on any queue sharing the
+supplied context. It is valid to call this method multiple times to create
+subsequent executable graphs. It is also valid to continue to add new nodes to
+the modifiable graph instance after calling this function. It is valid to
+finalize an empty graph instance with no recorded commands.
+
+Preconditions:
+
+* This member function is only available when the `command_graph` state is
+  `graph_state::modifiable`.
+
+Parameters:
+
+* `syclContext` - The context associated with the queues to which the
+  executable graph will be able to be submitted.
+
+Returns: A new executable graph object which can be submitted to a queue.
+
+Exceptions:
+
+* Throws synchronously with error code `invalid` if the graph contains a cycle.
+  A cycle may be introduced to the graph via a call to `make_edge()` that
+  creates a forward dependency.
```

Comment:
It would be much more convenient to the user if the exception was thrown as soon as the circular dependency was added (e.g. from the `make_edge` call).  Can you do that instead?

Reply:
I agree that detecting a cycle and throwing an exception from the `make_edge` call would be more user friendly.
It would be clear which dependency caused the cycle, and therefore can be easily avoided. In that case a small overhead for repetitive cycle detection might be acceptable.

My main concern is that the explicit API is not necessarily user facing, but potentially also being used by a framework/middleware that can exclude cycles by design (e.g. only allows forward dependencies). In that case any extra overhead would be unacceptable.

Reply:
> In that case any extra overhead would be unacceptable.

I agree that there is some point where the overhead is unacceptable, but it is too absolutist to say that the overhead must be zero.

If you are mostly worried about the framework use case, where the framework knows there won't be any cycles, we could introduce a property which disables the error checking.  This would allow those users to avoid the overhead when they know they don't need it.

My high level point is that reporting good errors is important.  In fact, we have an overall group goal to improve our error messages.  When we report a cycle, the error message should somehow tell the user which node(s) are causing the cycle, so they know where to look.

Reply:
I cam across this reading through old comments.  What is the current thinking about this one?  I do not see any property that controls error reporting of cycles in the current version.

Reply:
Property being added in https://github.com/reble/llvm/pull/128